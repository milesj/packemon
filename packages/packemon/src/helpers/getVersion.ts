import fs from 'fs';
import { PackageStructure } from '@boost/common';

export function getVersion(): string {
	return (
		JSON.parse(
			// @ts-expect-error Not typed for URL
			fs.readFileSync(new URL('../../package.json', import.meta.url), 'utf8'),
		) as PackageStructure
	).version;
}
