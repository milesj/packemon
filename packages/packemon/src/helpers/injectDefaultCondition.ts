/* eslint-disable no-param-reassign */

import { PackageExportConditions, PackageExports } from '../types';

export function injectDefaultCondition(exportMap: PackageExports) {
	Object.entries(exportMap).forEach(([path, conditions]) => {
		if (typeof conditions !== 'object') {
			return;
		}

		// Order from most to least down-leveled
		if (!conditions.default) {
			for (const key of [
				'browser',
				'react-native',
				'electron',
				'node',
				'require',
			] as PackageExportConditions[]) {
				if (conditions[key]) {
					conditions.default = conditions[key];
					delete conditions[key];
					break;
				}
			}
		}

		const keys = Object.keys(conditions);

		if (keys.length === 1) {
			exportMap[path] =
				keys[0] === 'default' ? conditions.default : conditions[keys[0] as PackageExportConditions];
		} else {
			exportMap[path] = conditions;
		}
	});
}
