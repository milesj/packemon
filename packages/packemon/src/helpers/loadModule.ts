import { createRequire } from 'module';

const modRequire = createRequire(import.meta.url);

export function loadModule(name: string, message: string): unknown {
	try {
		return modRequire(name);
	} catch {
		throw new Error(`${message} Please install with \`yarn add --dev ${name}\`.`);
	}
}
