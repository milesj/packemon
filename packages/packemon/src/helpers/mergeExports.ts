/* eslint-disable no-param-reassign */

import { PackageExportConditions, PackageExportPaths } from '../types';

export function mergeExports(
	prev: PackageExportPaths | string,
	base: PackageExportPaths | string,
	root: boolean,
): PackageExportPaths | undefined {
	const next = typeof base === 'string' ? { default: base } : base;

	if (typeof prev === 'string') {
		return next;
	}

	Object.entries(next).forEach(([origKey, nextValue]) => {
		const key = origKey as PackageExportConditions;
		const prevValue = prev[key];

		if (!prevValue) {
			prev[key] = nextValue;
		} else if (nextValue) {
			prev[key] = mergeExports(prevValue, nextValue, false);
		}
	});

	// Always include a default export for tooling to work correctly
	const keys = Object.keys(prev);

	if (root && !prev.default) {
		for (const key of [
			'react-native',
			'electron',
			'browser',
			'node',
			'import',
			'require',
		] as PackageExportConditions[]) {
			if (prev[key]) {
				if (keys.length === 1) {
					return prev[key] as PackageExportPaths;
				}

				prev.default = prev[key];
				delete prev[key];

				break;
			}
		}
	}

	return prev;
}
