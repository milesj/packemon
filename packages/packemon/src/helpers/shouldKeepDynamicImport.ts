import { Format, Platform } from '../types';

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#browser_compatibility
export function shouldKeepDynamicImport(platform: Platform, format: Format): boolean {
	if (format === 'umd') {
		return false;
	}

	switch (platform) {
		case 'node':
			// >= v13.2
			// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/import#browser_compatibility
			return true;
		case 'electron':
		case 'browser':
			// >= 2019
			// https://caniuse.com/es6-module-dynamic-import
			return true;
		case 'native':
			// >= RN 0.72
			// https://metrobundler.dev/docs/module-api/#import-dynamic-import
			return true;
		default:
			return false;
	}
}
