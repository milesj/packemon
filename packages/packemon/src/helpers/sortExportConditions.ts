import { PackageExportPaths } from '../types';

// https://nodejs.org/api/packages.html#conditional-exports
const WEIGHTS = {
	types: 10, // Types must always be first
	misc: 20,
	'node-addons': 30,
	node: 31,
	import: 40,
	require: 50,
	default: 100, // Default must be last
};

export function sortExportConditions<T extends PackageExportPaths | string>(paths: T): T {
	if (typeof paths === 'string') {
		return paths;
	}

	const pathsList: { weight: number; key: string; value: PackageExportPaths | string }[] = [];

	Object.entries(paths).forEach(([key, value]) => {
		pathsList.push({
			weight: key in WEIGHTS ? WEIGHTS[key as 'misc'] : WEIGHTS.misc,
			key,
			value: sortExportConditions(value),
		});
	});

	pathsList.sort((a, d) => {
		const diff = a.weight - d.weight;

		return diff === 0 ? a.key.localeCompare(d.key) : diff;
	});

	return Object.fromEntries(pathsList.map((path) => [path.key, path.value])) as T;
}
