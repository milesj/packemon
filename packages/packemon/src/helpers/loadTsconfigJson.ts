import { Path, type PortablePath } from '@boost/common';
import type { FileSystem } from '../FileSystem';
import type { TSConfigStructure } from '../types';
import { loadModule } from './loadModule';

const CACHE = new Map<Path, TSConfigStructure>();

export function loadTsconfigJson(
	path: PortablePath,
	fs: FileSystem,
): TSConfigStructure | undefined {
	const tsconfigJsonPath = Path.create(path);
	const tsconfig = CACHE.get(tsconfigJsonPath);

	if (tsconfig) {
		return tsconfig;
	}

	if (!tsconfigJsonPath.exists()) {
		return undefined;
	}

	const ts = loadModule(
		'typescript',
		'TypeScript is required for config loading.',
	) as typeof import('typescript');

	// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
	const { config, error } = ts.readConfigFile(tsconfigJsonPath.path(), fs.readFile);

	const host = {
		getCanonicalFileName: (fileName: string) => fileName,
		getCurrentDirectory: () => ts.sys.getCurrentDirectory(),
		getNewLine: () => ts.sys.newLine,
	};

	// istanbul ignore next
	if (error) {
		throw new Error(ts.formatDiagnostic(error, host));
	}

	const result = ts.parseJsonConfigFileContent(
		config,
		ts.sys,
		tsconfigJsonPath.parent().path(),
		{},
		tsconfigJsonPath.path(),
	);

	// istanbul ignore next
	if (result.errors.length > 0) {
		throw new Error(ts.formatDiagnostics(result.errors, host));
	}

	CACHE.set(tsconfigJsonPath, result);

	return result;
}
