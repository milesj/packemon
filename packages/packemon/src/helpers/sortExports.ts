import { PackageExports } from '../types';
import { sortExportConditions } from './sortExportConditions';

const WEIGHTS: Record<string, number> = {
	'./package.json': 0, // First
	'.': 100, // Last
};

export function sortExports(exportMap: PackageExports): PackageExports {
	const paths = Object.keys(exportMap);

	paths.sort((a, d) => {
		const diff = (WEIGHTS[a] ?? 10) - (WEIGHTS[d] ?? 10);

		return diff === 0 ? d.length - a.length : diff;
	});

	return Object.fromEntries(paths.map((path) => [path, sortExportConditions(exportMap[path])]));
}
