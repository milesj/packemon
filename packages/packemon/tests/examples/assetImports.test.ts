import { describe } from 'vitest';
import { testExampleOutput } from '../helpers';

describe('Assets', () => {
	testExampleOutput('asset-imports.ts', 'babel');
	testExampleOutput('asset-imports.ts', 'swc');
});
