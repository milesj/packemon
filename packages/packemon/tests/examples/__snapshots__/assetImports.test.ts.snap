// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Assets transforms example test case: browser-current-esm 1`] = `
Array [
  "esm/index-browser-current-esm.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: current, Format: esm
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-current-lib 1`] = `
Array [
  "lib/index-browser-current-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: current, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-current-umd 1`] = `
Array [
  "umd/index-browser-current-umd.js",
  "(function (global, factory) {
  if (typeof define === \\"function\\" && define.amd) {
    define([\\"../assets/styles-173c885b.css\\", \\"../assets/logo-073ed8f7.svg\\"], factory);
  } else if (typeof exports !== \\"undefined\\") {
    factory(require(\\"../assets/styles-173c885b.css\\"), require(\\"../assets/logo-073ed8f7.svg\\"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.styles173c885b, global.logo073ed8f7);
    global.examples = mod.exports;
  }
})(typeof globalThis !== \\"undefined\\" ? globalThis : typeof self !== \\"undefined\\" ? self : this, function (_styles173c885b, _logo073ed8f) {
  \\"use strict\\";

  // Bundled with Packemon: https://packemon.dev
  // Platform: browser, Support: current, Format: umd
  const json = {
    list: [\\"foo\\", \\"bar\\", \\"baz\\"],
    number: 123,
    string: \\"abc\\",
    object: {
      foo: true
    }
  };
  console.log(json);
});
",
]
`;

exports[`Assets transforms example test case: browser-experimental-esm 1`] = `
Array [
  "esm/index-browser-experimental-esm.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: experimental, Format: esm
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-experimental-lib 1`] = `
Array [
  "lib/index-browser-experimental-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: experimental, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-experimental-umd 1`] = `
Array [
  "umd/index-browser-experimental-umd.js",
  "(function (global, factory) {
  if (typeof define === \\"function\\" && define.amd) {
    define([\\"../assets/styles-173c885b.css\\", \\"../assets/logo-073ed8f7.svg\\"], factory);
  } else if (typeof exports !== \\"undefined\\") {
    factory(require(\\"../assets/styles-173c885b.css\\"), require(\\"../assets/logo-073ed8f7.svg\\"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.styles173c885b, global.logo073ed8f7);
    global.examples = mod.exports;
  }
})(typeof globalThis !== \\"undefined\\" ? globalThis : typeof self !== \\"undefined\\" ? self : this, function (_styles173c885b, _logo073ed8f) {
  \\"use strict\\";

  // Bundled with Packemon: https://packemon.dev
  // Platform: browser, Support: experimental, Format: umd
  const json = {
    list: [\\"foo\\", \\"bar\\", \\"baz\\"],
    number: 123,
    string: \\"abc\\",
    object: {
      foo: true
    }
  };
  console.log(json);
});
",
]
`;

exports[`Assets transforms example test case: browser-legacy-esm 1`] = `
Array [
  "esm/index-browser-legacy-esm.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: legacy, Format: esm
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
var json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-legacy-lib 1`] = `
Array [
  "lib/index-browser-legacy-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: legacy, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

var json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-legacy-umd 1`] = `
Array [
  "umd/index-browser-legacy-umd.js",
  "(function (global, factory) {
  if (typeof define === \\"function\\" && define.amd) {
    define([\\"../assets/styles-173c885b.css\\", \\"../assets/logo-073ed8f7.svg\\"], factory);
  } else if (typeof exports !== \\"undefined\\") {
    factory(require(\\"../assets/styles-173c885b.css\\"), require(\\"../assets/logo-073ed8f7.svg\\"));
  } else {
    var mod = {
      exports: {}
    };
    factory(global.styles173c885b, global.logo073ed8f7);
    global.examples = mod.exports;
  }
})(typeof globalThis !== \\"undefined\\" ? globalThis : typeof self !== \\"undefined\\" ? self : this, function (_styles173c885b, _logo073ed8f) {
  \\"use strict\\";

  // Bundled with Packemon: https://packemon.dev
  // Platform: browser, Support: legacy, Format: umd
  var json = {
    list: [\\"foo\\", \\"bar\\", \\"baz\\"],
    number: 123,
    string: \\"abc\\",
    object: {
      foo: true
    }
  };
  console.log(json);
});
",
]
`;

exports[`Assets transforms example test case: browser-stable-esm 1`] = `
Array [
  "esm/index-browser-stable-esm.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: stable, Format: esm
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: browser-stable-lib 1`] = `
Array [
  "lib/index-browser-stable-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: browser, Support: stable, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: native-current-lib 1`] = `
Array [
  "lib/index-native-current-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: native, Support: current, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: native-experimental-lib 1`] = `
Array [
  "lib/index-native-experimental-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: native, Support: experimental, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: native-stable-lib 1`] = `
Array [
  "lib/index-native-stable-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: native, Support: stable, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-current-cjs 1`] = `
Array [
  "cjs/index-node-current-cjs.cjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: current, Format: cjs
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-current-lib 1`] = `
Array [
  "lib/index-node-current-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: current, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-current-mjs 1`] = `
Array [
  "mjs/index-node-current-mjs.mjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: current, Format: mjs
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-experimental-lib 1`] = `
Array [
  "lib/index-node-experimental-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: experimental, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-experimental-mjs 1`] = `
Array [
  "mjs/index-node-experimental-mjs.mjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: experimental, Format: mjs
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-legacy-lib 1`] = `
Array [
  "lib/index-node-legacy-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: legacy, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

var json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-legacy-mjs 1`] = `
Array [
  "mjs/index-node-legacy-mjs.mjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: legacy, Format: mjs
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
var json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-stable-cjs 1`] = `
Array [
  "cjs/index-node-stable-cjs.cjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: stable, Format: cjs
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-stable-lib 1`] = `
Array [
  "lib/index-node-stable-lib.js",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: stable, Format: lib
'use strict';

require('../assets/styles-173c885b.css');

require('../assets/logo-073ed8f7.svg');

const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;

exports[`Assets transforms example test case: node-stable-mjs 1`] = `
Array [
  "mjs/index-node-stable-mjs.mjs",
  "// Bundled with Packemon: https://packemon.dev
// Platform: node, Support: stable, Format: mjs
import '../assets/styles-173c885b.css';
import '../assets/logo-073ed8f7.svg';
const json = {
  list: [\\"foo\\", \\"bar\\", \\"baz\\"],
  number: 123,
  string: \\"abc\\",
  object: {
    foo: true
  }
};
console.log(json);
",
]
`;
